File: backend/app/services/m3u_service.py
import aiohttp
import logging
from typing import List, Optional
from ..models import Channel
import re

logger = logging.getLogger(__name__)


class M3UService:
    async def download_and_parse(self, url: str) -> List[Channel]:
        """Download M3U file and parse it into channels"""
        logger.info(f"Downloading M3U from {url}")

        try:
            async with aiohttp.ClientSession() as session:
                async with session.get(url) as response:
                    if not response.ok:
                        raise Exception(f"HTTP {response.status}: {response.reason}")

                    content = await response.text()
                    logger.info(f"Downloaded {len(content)} bytes")

                    return self.parse_m3u(content)

        except Exception as e:
            logger.error(f"Failed to download M3U: {str(e)}")
            raise

    def parse_m3u(self, content: str) -> List[Channel]:
        """Parse M3U content into channel objects"""
        lines = content.split("\n")
        channels = []
        current_channel = None

        for line in lines:
            line = line.strip()

            if line.startswith("#EXTINF:"):
                # Parse channel info
                info = line[8:].split(",", 1)
                attrs = self._parse_attributes(info[0])
                name = (
                    info[1].strip()
                    if len(info) > 1
                    else attrs.get("tvg-name", "Unknown")
                )

                current_channel = {
                    "id": attrs.get("tvg-id", f"channel-{len(channels)}"),
                    "name": name,
                    "group": attrs.get("group-title", "Uncategorized"),
                    "logo": attrs.get("tvg-logo"),
                    "is_favorite": False,
                }

            elif line.startswith("http") and current_channel:
                current_channel["url"] = line
                channels.append(current_channel)
                current_channel = None

        logger.info(f"Parsed {len(channels)} channels")
        return channels

    def _parse_attributes(self, attr_string: str) -> dict:
        """Parse EXTINF attributes into a dictionary"""
        attrs = {}
        matches = re.finditer(r'([a-zA-Z-]+)="([^"]*)"', attr_string)
        for match in matches:
            key, value = match.groups()
            attrs[key] = value
        return attrs

File: frontend/src/services/m3uService.ts
import { Channel } from '../models/Channel';
import { channelService } from './channelService';

export {};

interface ProgressCallback {
  (loaded: number, total: number): void;
}

const logStyles = {
  info: 'color: #4CAF50; font-weight: bold',  // Green
  error: 'color: #f44336; font-weight: bold', // Red
  warn: 'color: #ff9800; font-weight: bold',  // Orange
};

export const m3uService = {
  async parseM3U(url: string, onProgress?: ProgressCallback): Promise<Channel[]> {
    try {
      console.log('%c[M3U] Starting download from %s', logStyles.info, url);
      const response = await fetch(url);
      
      if (!response.ok) {
        console.error(
          '%c[M3U] HTTP error: %s %s', 
          logStyles.error,
          response.status,
          response.statusText
        );
        throw new Error(`HTTP error: ${response.status} ${response.statusText}`);
      }

      const reader = response.body?.getReader();
      const contentLength = +(response.headers.get('Content-Length') ?? '0');
      
      if (!reader) {
        throw new Error('Failed to get response reader');
      }

      let receivedLength = 0;
      let chunks: Uint8Array[] = [];

      while (true) {
        const { done, value } = await reader.read();
        
        if (done) break;
        
        chunks.push(value);
        receivedLength += value.length;
        
        if (onProgress) {
          onProgress(receivedLength, contentLength);
        }
      }

      const chunksAll = new Uint8Array(receivedLength);
      let position = 0;
      for (const chunk of chunks) {
        chunksAll.set(chunk, position);
        position += chunk.length;
      }

      const text = new TextDecoder('utf-8').decode(chunksAll);
      console.log(
        '%c[M3U] Downloaded %sMB', 
        logStyles.info,
        (receivedLength / 1024 / 1024).toFixed(2)
      );

      const lines = text.split('\n');
      console.log(`[M3U] Parsing ${lines.length} lines`);
      
      const channels: Channel[] = [];
      let currentChannel: Partial<Channel> = {
        channel_number: 0,
        guide_id: '',
        name: '',
        group: 'Default',
        isFavorite: false
      };

      lines.forEach((line, index) => {
        line = line.trim();
        
        if (line.startsWith('#EXTINF:')) {
          // Parse channel info
          const info = line.substring(8).split(',');
          const attributes = info[0].match(/([a-zA-Z-]+)="([^"]*)"/g);
          
          currentChannel = {
            channel_number: index + 1,
            guide_id: `channel-${index}`,
            name: info[1].trim(),
            group: 'Default',
            isFavorite: false
          };

          // Parse attributes
          attributes?.forEach(attr => {
            const [key, value] = attr.split('=');
            const cleanValue = value.replace(/"/g, '');
            
            switch (key) {
              case 'tvg-logo':
                currentChannel.logo = cleanValue;
                break;
              case 'group-title':
                currentChannel.group = cleanValue;
                break;
              case 'tvg-id':
                currentChannel.guide_id = cleanValue;
                break;
              case 'tvg-name':
                currentChannel.name = cleanValue || currentChannel.name;
                break;
            }
          });
        } else if (line.startsWith('http')) {
          currentChannel.url = line;
          if (currentChannel.name && currentChannel.url) {
            channels.push(currentChannel as Channel);
          }
          currentChannel = {};
        }
      });

      console.log(`[M3U] Found ${channels.length} channels in ${Object.keys(
        channels.reduce((acc, c) => ({ ...acc, [c.group]: true }), {})
      ).length} groups`);

      console.log('[M3U] Saving channels to database...');
      await channelService.saveChannels(channels);
      console.log('[M3U] Channels saved successfully');

      return channels;
    } catch (error: unknown) {
      const errorMessage = error instanceof Error ? error.message : 'Unknown error';
      console.error('%c[M3U] Error details:', logStyles.error, {
        message: errorMessage,
        error,
        url
      });
      throw error;
    }
  }
}; 